README.md
# Lockit SQL adapter

[![Build Status](https://travis-ci.org/zemirco/lockit-sql-adapter.svg?branch=master)](https://travis-ci.org/zemirco/lockit-sql-adapter) [![NPM version](https://badge.fury.io/js/lockit-sql-adapter.svg)](http://badge.fury.io/js/lockit-sql-adapter)

SQL adapter for [Lockit](https://github.com/zemirco/lockit).

## Installation

`npm install lockit-sql-adapter`

```js
var adapter = require('lockit-sql-adapter');
```

The adapter is built on top of [sequelize](http://sequelizejs.com/).
The following databases are supported:

 - MySQL
 - MariaDB (not yet tested but should work)
 - SQLite
 - PostgreSQL

You have to install the connector for your database of choice manually.

```
npm install pg       # for postgres
npm install mysql    # for mysql
npm install sqlite3  # for sqlite
npm install mariasql # for mariasql
```

## Configuration

The following settings are required.

```js
// for postgres
exports.db = {
  url: 'postgres://127.0.0.1:5432/',
  name: 'users',
  collection: 'my_user_table'  // table name
};

// for mysql
// exports.db = {
//   url: 'mysql://127.0.0.1:3306/',
//   name: 'users',
//   collection: 'my_user_table'  // table name
// };

// for sqlite
// exports.db = {
//   url: 'sqlite://',
//   name: ':memory:',
//   collection: 'my_user_table'  // table name
// };
```

## Features

### 1. Create user

`adapter.save(name, email, pass, callback)`

 - `name`: String - i.e. 'john'
 - `email`: String - i.e. 'john@email.com'
 - `pass`: String - i.e. 'password123'
 - `callback`: Function - `callback(err, user)` where `user` is the new user now in our database.

The `user` object has the following properties

 - `_id`: unique id
 - `name`: username chosen during sign up
 - `email`: email that was provided at the beginning
 - `salt`: salt generated by `crypto.randomBytes()`
 - `derived_key`: password hash generated by pbkdf2
 - `signupTimestamp`: Date object to remember when the user signed up
 - `signupToken`: unique token sent to user's email for email verification
 - `signupTokenExpires`: Date object usually 24h ahead of `signupTimestamp`
 - `failedLoginAttempts`: save failed login attempts during login process, default is `0`

```js
adapter.save('john', 'john@email.com', 'secret', function(err, user) {
  if (err) console.log(err);
  console.log(user);
  // {
  //   _id: 1,
  //   name: 'john',
  //   email: 'john@email.com',
  //   derived_key: 'c4c7a83f7b3936437798316d4c7b8c7b731a55dc',
  //   salt: 'ff449a4980a58a80c4ed80bddd34b8c9',
  //   signupToken: '13eefbe7-6bc8-43f5-b27f-0bf0ca98b8db',
  //   signupTimestamp: Fri Apr 11 2014 21:37:47 GMT+0200 (CEST),
  //   signupTokenExpires: Sat Apr 12 2014 21:37:47 GMT+0200 (CEST),
  //   failedLoginAttempts: 0,
  //   emailVerificationTimestamp: null,
  //   emailVerified: null,
  //   pwdResetToken: null,
  //   pwdResetTokenExpires: null,
  //   accountLocked: null,
  //   accountLockedUntil: null,
  //   previousLoginTime: null,
  //   previousLoginIp: null,
  //   currentLoginTime: null,
  //   currentLoginIp: null
  // }
});
```

### 2. Find user

`adapter.find(match, query, callback)`

 - `match`: String - one of the following: 'name', 'email' or 'signupToken'
 - `query`: String - corresponds to `match`, i.e. 'john@email.com'
 - `callback`:  Function - `callback(err, user)`

```js
adapter.find('name', 'john', function(err, user) {
  if (err) console.log(err);
  console.log(user);
  // {
  //   _id: 1,
  //   name: 'john',
  //   email: 'john@email.com',
  //   derived_key: '75b43d8393715cbf476ee55b12f888246d7f7015',
  //   salt: 'f39f9a5104e5ae61347dced750b63b16',
  //   signupToken: '6c93c6f8-06b6-4c6d-be58-1e89e8590d0f',
  //   signupTimestamp: Fri Apr 11 2014 21:39:28 GMT+0200 (CEST),
  //   signupTokenExpires: Sat Apr 12 2014 21:39:28 GMT+0200 (CEST),
  //   failedLoginAttempts: 0,
  //   emailVerificationTimestamp: null,
  //   emailVerified: null,
  //   pwdResetToken: null,
  //   pwdResetTokenExpires: null,
  //   accountLocked: null,
  //   accountLockedUntil: null,
  //   previousLoginTime: null,
  //   previousLoginIp: null,
  //   currentLoginTime: null,
  //   currentLoginIp: null
  // }
});
```

### 3. Update user

`adapter.update(user, callback)`

 - `user`: Object - must have `_id` key
 - `callback`: Function - `callback(err, user)` - `user` is the updated user object

```js
// get a user from db first
adapter.find('name', 'john', function(err, user) {
  if (err) console.log(err);

  // add some new properties to our existing user
  user.firstOldKey = 'and some value';
  user.secondOldKey = true;

  // save updated user to db
  adapter.update(user, function(err, user) {
    if (err) console.log(err);
    // ...
  });
});
```

### 4. Delete user

`adapter.remove(name, callback)`

 - `name`: String
 - `callback`: Function - `callback(err, res)` - `res` is `true` if everything went fine

```js
adapter.remove('john', function(err, res) {
  if (err) console.log(err);
  console.log(res);
  // true
});
```

## Test

`grunt`

## License

MIT


History.md

##### 0.5.3 / 2015-07-06

- fix eslint errors in test files

##### 0.5.2 / 2015-07-03

- use make instead grunt
- update dependencies
- use eslint

##### 0.5.1 / 2014-05-19

- code refactoring
- update dependencies

##### 0.5.0 / 2014-04-17

- database config has to be an Object and the database name isn't part of the
  connection string anymore

  **old**

  ```js
  exports.db = 'mysql://127.0.0.1:9821/users';
  exports.dbCollection = 'my_user_table';
  ```

  **new**

  ```js
  exports.db = {
    url: 'mysql://127.0.0.1:3306/',
    name: 'users',
    collection: 'my_user_table'
  },
  ```

##### 0.4.0 / 2014-04-10

- `username` becomes `name`
- use built-in [pbkdf2](http://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2_password_salt_iterations_keylen_callback)
  instead [bcrypt](https://github.com/ncb000gt/node.bcrypt.js/)

##### 0.3.0 / 2014-04-03

- simplify `remove()` method

  ```js
  adapter.remove(match, query, callback)
  ```

  becomes

  ```js
  adapter.remove(username, callback)
  ```

...

##### 0.0.1 / 2014-01-19

 - initial release


test/test.js
'use strict';

var should = require('should');
var config = require('./config.js');

// run the same tests for postgres, mysql and sqlite
Object.keys(config.dbs).forEach(function(obj) {

  var conf = {
    db: {
      url: config.dbs[obj].url,
      name: config.dbs[obj].name,
      collection: config.dbs[obj].collection
    },
    signup: config.signup
  };

  var adapter = require('../index.js')(conf);

  // start the tests
  describe(obj + ' adapter for lockit', function() {

    // needed later for test 'find user by signup token'
    var _tmp_signupToken = '';

    it('should create a new user', function(done) {
      adapter.save('john', 'john@email.com', 'secret', function(err, res) {
        if (err) {console.log(err); }
        res.should.have.property('signupToken');
        res.signupToken.should.match(/[0-9a-f]{22}|[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/);
        res.should.have.property('failedLoginAttempts');
        res.failedLoginAttempts.should.equal(0);
        _tmp_signupToken = res.signupToken;
        res.email.should.equal('john@email.com');
        done();
      });
    });

    it('should find a user by name', function(done) {
      adapter.find('name', 'john', function(err, res) {
        if (err) {console.log(err); }
        res.name.should.equal('john');
        res.email.should.equal('john@email.com');
        done();
      });
    });

    it('should return undefined when no user is found', function(done) {
      adapter.find('name', 'jim', function(err, res) {
        if (err) {console.log(err); }
        should.not.exist(err);
        should.not.exist(res);
        done();
      });
    });

    it('should find a user by email', function(done) {
      adapter.find('email', 'john@email.com', function(err, res) {
        if (err) {console.log(err); }
        res.name.should.equal('john');
        res.email.should.equal('john@email.com');
        done();
      });
    });

    it('should find a user by signup token', function(done) {
      adapter.find('signupToken', _tmp_signupToken, function(err, res) {
        if (err) {console.log(err); }
        res.name.should.equal('john');
        res.email.should.equal('john@email.com');
        done();
      });
    });

    it('should update an existing user', function(done) {
      adapter.find('name', 'john', function(err, doc) {
        if (err) {console.log(err); }
        doc.signupToken = 'works';
        doc.accountLocked = true;
        adapter.update(doc, function(updateErr, res) {
          if (updateErr) {console.log(updateErr); }
          res.signupToken.should.equal('works');
          res.accountLocked.should.be.true;
          done();
        });
      });
    });

    it('should remove a user', function(done) {
      adapter.save('jeff', 'jeff@email.com', 'secret', function(err) {
        if (err) {console.log(err); }
        adapter.remove('jeff', function(removeErr, res) {
          if (removeErr) {console.log(removeErr); }
          res.should.be.true;
          done();
        });
      });
    });

    it('should return an error when remove cannot find a user', function(done) {
      adapter.remove('steve', function(err) {
        err.message.should.equal('lockit - Cannot find user "steve"');
        done();
      });
    });

  });

  // remove users db
  after(function(done) {
    adapter.remove('john', done);
  });
});


test/config.js
exports.dbs = {
  postgres: {
    url: 'postgres://postgres:@127.0.0.1:5432/',
    name: 'users',
    collection: 'my_user_table'
  },
  mysql: {
    url: 'mysql://travis:@127.0.0.1:3306/',
    name: 'users',
    collection: 'my_user_table'
  },
  sqlite: {
    url: 'sqlite://',
    name: ':memory:',
    collection: 'my_user_table'
  }
};

exports.signup = {
  tokenExpiration: '1 day'
};


index.js
'use strict';

var uuid = require('uuid');
var pwd = require('couch-pwd');
var ms = require('ms');
var moment = require('moment');
var Sequelize = require('sequelize');

/**
 * Adapter constructor function
 *
 * @example
   var Adapter = require('lockit-sql-adapter');
   var config = require('./config.js');
   var adapter = new Adapter(config);
 *
 * @param {Object} config
 * @constructor
 */
var Adapter = module.exports = function(config) {

  if (!(this instanceof Adapter)) {return new Adapter(config); }

  this.config = config;

  // create connection string
  var uri = config.db.url + config.db.name;
  var sequelize = new Sequelize(uri, {
    storage: config.db.name
  });

  this.User = sequelize.define('User', {
    // make id like CouchDB and MongoDB
    _id: {
      type: Sequelize.INTEGER,
      autoIncrement: true,
      primaryKey: true
    },
    // signup
    name: Sequelize.STRING,
    email: Sequelize.STRING,
    derived_key: Sequelize.STRING,
    salt: Sequelize.STRING,
    signupToken: Sequelize.STRING,
    signupTimestamp: Sequelize.DATE,
    signupTokenExpires: Sequelize.DATE,
    failedLoginAttempts: Sequelize.INTEGER,
    emailVerificationTimestamp: Sequelize.DATE,
    emailVerified: Sequelize.BOOLEAN,
    // forgot password
    pwdResetToken: Sequelize.STRING,
    pwdResetTokenExpires: Sequelize.DATE,
    // login
    accountLocked: Sequelize.BOOLEAN,
    accountLockedUntil: Sequelize.DATE,
    previousLoginTime: Sequelize.DATE,
    previousLoginIp: Sequelize.STRING,
    currentLoginTime: Sequelize.DATE,
    currentLoginIp: Sequelize.STRING
  }, {
    tableName: config.db.collection,   // this will define the table's name
    timestamps: false                 // this will deactivate the timestamp columns
  });

  this.User.sequelize.sync({})
    .then(function() {
      // you can now use User to create new instances
    })
    .error(function(err) {
      throw (err);
    });

};


/**
 * Create a new user and save it to db.
 *
 * @example
   adapter.save('john', 'john@email.com', 'secret', function(err, user) {
     if (err) console.log(err);
     console.log(user);
     // {
     //   _id: 1,
     //   name: 'john',
     //   email: 'john@email.com',
     //   derived_key: 'c4c7a83f7b3936437798316d4c7b8c7b731a55dc',
     //   salt: 'ff449a4980a58a80c4ed80bddd34b8c9',
     //   signupToken: '13eefbe7-6bc8-43f5-b27f-0bf0ca98b8db',
     //   signupTimestamp: Fri Apr 11 2014 21:37:47 GMT+0200 (CEST),
     //   signupTokenExpires: Sat Apr 12 2014 21:37:47 GMT+0200 (CEST),
     //   failedLoginAttempts: 0,
     //   emailVerificationTimestamp: null,
     //   emailVerified: null,
     //   pwdResetToken: null,
     //   pwdResetTokenExpires: null,
     //   accountLocked: null,
     //   accountLockedUntil: null,
     //   previousLoginTime: null,
     //   previousLoginIp: null,
     //   currentLoginTime: null,
     //   currentLoginIp: null
     // }
   });
 *
 * @param {String} name - User name
 * @param {String} email - User email
 * @param {String} pw - Plain text password
 * @param {Function} done - Callback function having `err` and `user` as arguments
 */
Adapter.prototype.save = function(name, email, pw, done) {
  var that = this;
  var now = moment().toDate();
  var timespan = ms(that.config.signup.tokenExpiration);
  var future = moment().add(timespan, 'ms').toDate();
  // create hashed password
  pwd.hash(pw, function(err, salt, hash) {
    if (err) {return done(err); }
    var user = that.User.build({
      name: name,
      email: email,
      signupToken: uuid.v4(),
      signupTimestamp: now,
      signupTokenExpires: future,
      failedLoginAttempts: 0,
      salt: salt,
      derived_key: hash
    });
    // save user to db
    user.save()
      .then(function() {
        // find user to return it in callback
        that.User.find({ where: {email: email} })
          .then(function(foundUser) {
            done(null, foundUser.dataValues);
          })
          .error(function(findErr) {
            done(findErr);
          });
      })
      .error(function(saveErr) {
        done(saveErr);
      });
  });
};

/**
 * Find a user in db.
 *
 * @example
   adapter.find('name', 'john', function(err, user) {
     if (err) console.log(err);
     console.log(user);
     // {
     //   _id: 1,
     //   name: 'john',
     //   email: 'john@email.com',
     //   derived_key: '75b43d8393715cbf476ee55b12f888246d7f7015',
     //   salt: 'f39f9a5104e5ae61347dced750b63b16',
     //   signupToken: '6c93c6f8-06b6-4c6d-be58-1e89e8590d0f',
     //   signupTimestamp: Fri Apr 11 2014 21:39:28 GMT+0200 (CEST),
     //   signupTokenExpires: Sat Apr 12 2014 21:39:28 GMT+0200 (CEST),
     //   failedLoginAttempts: 0,
     //   emailVerificationTimestamp: null,
     //   emailVerified: null,
     //   pwdResetToken: null,
     //   pwdResetTokenExpires: null,
     //   accountLocked: null,
     //   accountLockedUntil: null,
     //   previousLoginTime: null,
     //   previousLoginIp: null,
     //   currentLoginTime: null,
     //   currentLoginIp: null
     // }
   });
 *
 * @param {String} match - The key to look for. Can be `'name'`, `'email'` or `'signupToken'`
 * @param {String} query - The corresponding value for `match`
 * @param {Function} done - Callback function having `err` and `user` as arguments
 */
Adapter.prototype.find = function(match, query, done) {
  var qry = {};
  qry[match] = query;
  this.User.find({ where: qry })
    .then(function(user) {
      // create empty object in case no user is found
      user = user || {};
      done(null, user.dataValues);
    })
    .error(function(err) {
      done(err);
    });
};

/**
 * Update an already existing user in db.
 *
 * @example
   // get a user from db first
   adapter.find('name', 'john', function(err, user) {
     if (err) console.log(err);

     // add some new properties to our existing user
     user.firstOldKey = 'and some value';
     user.secondOldKey = true;

     // save updated user to db
     adapter.update(user, function(err, user) {
       if (err) console.log(err);
       // ...
     });
   });
 *
 * @param {Object} user - User object with `user._id`
 * @param {Function} done - Callback function having `err` and `user` as arguments
 */
Adapter.prototype.update = function(user, done) {
  var that = this;
  that.User.update(user, {where: {_id: user._id}})
    .then(function() {
      that.User.findById(user._id)
        .then(function(foundUser) {
          done(null, foundUser.dataValues);
        })
        .error(function(err) {
          done(err);
        });
    })
    .error(function(err) {
      done(err);
    });
};

/**
 * Remove an existing user from db.
 *
 * @example
   adapter.remove('john', function(err, res) {
     if (err) console.log(err);
     console.log(res);
     // true
   });
 *
 * @param {String} name - Username
 * @param {Function} done - Callback function having `err` and `res` arguments
 */
Adapter.prototype.remove = function(name, done) {
  this.User.find({ where: {name: name} })
    .then(function(user) {
      if (!user) {return done(new Error('lockit - Cannot find user "' + name + '"')); }
      user.destroy()
        .then(function() {
          done(null, true);
        })
        .error(function(err) {
          done(err);
        });
    })
    .error(function(err) {
      done(err);
    });
};
